### DISCUSSION TITLE
Can/should I use both `-Xsource:3` and `-source:3.0-migration` in a cross-compiled codebase?

### DISCUSSION THREAD

**Discussion by armanbilge** [scala/scala3#17591] [General Question] [answered]
I'd like to refresh the compiler flags configured by [sbt-typelevel](https://typelevel.org/sbt-typelevel/) which is used to cross-compile several libraries for both Scala 2 and 3. I have a question about:

- `-Xsource:3` in Scala 2, which "enables some Scala 3 syntax and behavior"

- `-source:3.0-migration` in Scala 3, which "accepts some of the old Scala 2.13 syntax"

Ref: https://docs.scala-lang.org/scala3/guides/migration/tooling-tour.html

Does it make sense to use both of these flags in a cross-compiled codebase? I ask, because the former sounds like "make Scala 2 accept Scala 3-ish sources" and the latter sounds like "make Scala 3 accept Scala 2-ish sources". So I wonder if we should pick one strategy (aim for either Scala 2-ish or Scala 3-ish sources) and use the one flag that makes sense for that.

**_Also_** what about `-source:3.3-migration`? 😅 Can _that_ be mixed with `-Xsource:3`?

### Case Study

Consider the following program:

```scala
case class Foo private (bar: String)

object Main {
  def main(args: Array[String]): Unit =
    println(Foo("bar"))
}
```

Actually, this doesn't compile out-of-the-box with Scala 3.3.0.
```
[error] object Foo does not take parameters
[error]     println(Foo("bar"))
[error]             ^^^
```

Now here is what happens with various Scala version and compiler flag configurations:

| Options / Scala         | 2.13.10 | 3.3.0 |
|-------------------------|---------|-------|
| no options              | ✅       | ❌     |
| `-Xsource:3`            | ❌       | n/a   |
| `-source:3.0-migration` | n/a     | ✅     |
| `-source:3.3-migration` | n/a     | ❌     |

#### Analysis:
- `-Xsource:3` and `-source:3.0-migration` have the _opposite_ effect on their respective compilers. So using them together seems pretty dumb.
- `-Xsource:3` and `-source:3.3-migration` in this specific situation have the same result. So maybe they can be used together?

**Comment by som-snytt** [comment]
There is an old ticket somewhere to document `-Xsource` for Scala 2.

The purpose of `-Xsource:3` was to get source in a place where it compiles straight under scala 3.

**Comment by smarter** [comment]
I don't think a plugin should unilaterally touch -source, since it can lead to a lot of confusion down the road without the user realizing what the culprit is. For example -source:3.0-migration can break Scala 3 code because it changes implicit resolution rules (e.g., nested implicits will lead to ambiguity errors, but Scala 3 APIs using context function types routinely rely on that).

**Reply by armanbilge** [reply]
What do you mean by unilaterally? The goal here is to ship a reasonable default (that can be user-overridden) for hundreds of projects that are trying to cross-compile for Scala 2 and 3. If a project is not cross-compiling, I agree that we should not add these flags (and indeed we don't currently).

Is there no reasonable default? 😅

**Reply by smarter** [reply]
(Scala 2's -Xsource:3 is much more benign so I wouldn't be worried too much about that one, but it only really brings value if one wants to start using new syntactic forms (like `foo(s*)` instead of `foo(s: _*)`) in their codebase)

**Reply by armanbilge** [reply]
> Scala 2's -Xsource:3 is much more benign so I wouldn't be worried too much about that one, but it only really brings value if one wants to start using new syntactic forms (like `foo(s*)` instead of `foo(s: _*)`) in their codebase)

I'm not sure I agree ... or perhaps we have different definitions of "benign". I would understand if it only enabled new syntactic forms, but in my example above it actually broke existing code. That's hardly benign IMO.

**Reply by smarter** [reply]
> What do you mean by unilaterally? 

Without the project author explicitly opting into it. I think -source is dangerous and should not be enabled without thinking through all the consequences (and ideally it should only be used temporarily during a migration)

**Reply by armanbilge** [reply]
> and ideally it should only be used temporarily during a migration

Hmm, you mean like migrating from Scala 2 to Scala 3? What about for libraries that are permanently cross-compiled?

**Reply by smarter** [reply]
Such libraries need to be written in the common subset of Scala 2 and 3 which has the same meaning in both languages (and use version-specific sources when that's not possible).

**Reply by armanbilge** [reply]
Ok, thanks! So if I'm understanding correctly, cross-compiling libraries should use _neither_ `-Xsource:3` nor `-source:3.x-migration` ?

**Reply by smarter** [reply, marked as answer]
If you're cross-compiling, then I think -Xsource:3 isn't a problem because the things it reject will also be rejected by Scala 3 anyway, and it will mean you can use new syntax, but it's not a big deal either way.

**Reply by som-snytt** [reply]
I disagree that `-Xsource:3` doesn't matter. Modulo bugs, you can use leading infix. You can't use early defs. It was a bigger commitment when it included the implicits resolution change.
